crates/biome_js_semantic/src/events.rs

```diff
@@ -158,44 +152,84 @@ impl SemanticEvent {
 /// ```
 #[derive(Default, Debug)]
 pub struct SemanticEventExtractor {
+    // Event queue
     stash: VecDeque<SemanticEvent>,
+    /// Stack of scopes
     scopes: Vec<Scope>,
-    next_scope_id: usize,
-    /// At any point this is the set of available bindings and
-    /// the range of its declaration
+    // Number of generated scopes
+    // This allows assigning a unique scope id for every scope.
+    scope_count: usize,
+    /// At any point this is the set of available bindings in the current scope
     bindings: FxHashMap<BindingName, BindingInfo>,
+    // Type parameters bound in a `infer T` clause.
     infers: Vec<TsTypeParameterName>,
 }
 
-/// Holds the text range of the token when it is bound,
-/// along with the kind of declaration
-#[derive(Debug)]
-struct BindingInfo {
-    text_range: TextRange,
-    /// For export determination, it is necessary to provide
-    /// information on how a specific token is bound
-    declaration_kind: JsSyntaxKind,
+/// A declaration is either a type, a value, or both.
+///
+/// For example a class is botha  type and a value in TypeScript.
+/// A variable declaration is always a value, and a type alias is always a type.
+#[derive(Debug, Hash, Eq, PartialEq, Copy, Clone)]
+enum DeclarationKind {
+    Type,
+    Value,
+    Both,
 }
 
+/// A binding name is either a type or a value.
+///
+/// If a declaration is both a type and a value,
+/// then two bindings are associated to the declaration a type and a value bindings.
+///
+/// This allows to properly detect type and value shadowing in inner scopes.
 #[derive(Debug, Hash, Eq, PartialEq, Clone)]
-struct BindingName {
-    name: TokenText,
+enum BindingName {
+    Type(TokenText),
+    Value(TokenText),
 }
 
+impl BindingName {
+    /// Returns a binding with the same name, but the other kind.
+    fn dual(self) -> Self {
+        match self {
+            Self::Type(name) => Self::Value(name),
+            Self::Value(name) => Self::Type(name),
+        }
+    }
+}
+
+/// Holds the text range of the token when it is bound,
+/// along with the kind of declaration
 #[derive(Debug)]
+struct BindingInfo {
+    /// Range of the identifier binding.
+    name_range: TextRange,
+    /// For export determination, it is necessary to provide information on how a specific token is bound.
+    declaration_syntax_kind: JsSyntaxKind,
+}
+
+/// This type allows reporting a reference and bind to a binding (if any) later.
+#[derive(Debug, Clone)]
 enum Reference {
-    Read { range: TextRange, is_exported: bool },
-    Write { range: TextRange },
+    /// Read and export a type, a value, or both.
```

Comment:
```suggestion
    /// Read and export a type, a value, or both.
    /// ```js
    /// export { a }
    /// export type { b }
    /// ```
```

Some example is always helpful